src/core/compile.ml

```diff
@@ -109,13 +109,46 @@ let compile_file : file_path -> Sign.t = fun fname ->
 let _ =
   let require mp =
     (* Save the current console state. *)
-    Console.push_state ();
+    State.push ();
     (* Restore the console state to default for compiling. *)
     Console.reset_default ();
     (* Compile and go back to previous state. *)
     try
       ignore (compile false mp);
-      try Console.pop_state () with _ -> assert false (* Unreachable. *)
-    with e -> Console.pop_state (); raise e
+      try State.pop () with _ -> assert false (* Unreachable. *)
+    with e -> State.pop (); raise e
   in
   Stdlib.(Parser.require := require)
+
+(** Pure wrappers around compilation functions. Functions provided perform the
+    same computations as the ones defined earlier, but restores the state when
+    they have finished. An optional library mapping or state can be passed as
+    argument to change the settings. *)
+module Pure : sig
+  val compile : ?lm:string -> ?st:State.t -> bool -> Path.t -> Sign.t
+  val compile_file : ?lm:string -> ?st:State.t -> file_path -> Sign.t
+end = struct
+
+  (* [pure_apply_cfg ?lm ?st f] is function [f] but pure (without side
+     effects).  The side effects taken into account occur in
+     {!val:Console.State.t}, {!val:Files.lib_mappings} and in the meta
+     variable counter {!module:Terms.Meta}. Arguments [?lm] allows to set the
+     libray mappings and [?st] sets the state. *)
+  let pure_apply_cfg : ?lm:string -> ?st:State.t -> ('a -> 'b) -> 'a -> 'b =
+    fun ?lm ?st f x ->
+    let libmap = !lib_mappings in
+    State.push ();
+    Option.iter new_lib_mapping lm;
+    Option.iter State.apply st;
+    let res = f x in
+    State.pop ();
+    Terms.Meta.reset_key_counter ();
```

Comment:
Ideally, this counter should be reset at the beginning of each top-level command, so in handle_cmd when proof_data = None. Could this be done?

Reply:
Are you happy with 694e8f4?